sdk/python/feast/infra/offline_stores/remote.py

```diff
@@ -185,5 +258,152 @@ def offline_write_batch(
         table: pyarrow.Table,
         progress: Optional[Callable[[int], Any]],
     ):
-        # TODO Implementation here.
-        raise NotImplementedError
+        assert isinstance(config.offline_store, RemoteOfflineStoreConfig)
+
+        # Initialize the client connection
+        location = f"grpc://{config.offline_store.host}:{config.offline_store.port}"
+        client = fl.connect(location=location)
+        logger.info(f"Connecting FlightClient at {location}")
+
+        feature_view_names = [feature_view.name]
+        name_aliases = [feature_view.projection.name_alias]
+
+        api_parameters = {
+            "feature_view_names": feature_view_names,
+            "progress": progress,
+            "name_aliases": name_aliases,
+        }
+
+        command_descriptor = _call_put(
+            api="offline_write_batch",
+            api_parameters=api_parameters,
+            client=client,
+            table=table,
+            entity_df=None,
+        )
+        bytes = command_descriptor.serialize()
+
+        client.do_action(pa.flight.Action("offline_write_batch", bytes))
+
+
+def _create_retrieval_metadata(feature_refs: List[str], entity_df: pd.DataFrame):
+    entity_schema = _get_entity_schema(
+        entity_df=entity_df,
+    )
+
+    event_timestamp_col = offline_utils.infer_event_timestamp_from_entity_df(
+        entity_schema=entity_schema,
+    )
+
+    timestamp_range = _get_entity_df_event_timestamp_range(
+        entity_df, event_timestamp_col
+    )
+
+    return RetrievalMetadata(
+        features=feature_refs,
+        keys=list(set(entity_df.columns) - {event_timestamp_col}),
+        min_event_timestamp=timestamp_range[0],
+        max_event_timestamp=timestamp_range[1],
+    )
+
+
+def _get_entity_schema(entity_df: pd.DataFrame) -> Dict[str, np.dtype]:
```

Comment:
I think `entity_df` is of type `Union[pd.DataFrame, str]`, what happens in case of `str`?

Reply:
I actually took those method from the ibis implementation and there are ignoring the str option as well and in the file implementation they are doing:
```        
if not isinstance(entity_df, pd.DataFrame) and not isinstance(entity_df, dd.DataFrame):
            raise ValueError(
                f"Please provide an entity_df of type {type(pd.DataFrame)} instead of type {type(entity_df)}"
            )
```
Should I also enforce it to be DataFrame?

Reply:
I see that in another function you instead assume it's a DataFrame, you can probably do the same here.